Cabal/src/Distribution/PackageDescription/Check.hs

```diff
@@ -2350,7 +2351,7 @@ isGoodRelativeDirectoryPath = state0
 --     | x <= COtherReserved -> 6
 --     | otherwise           -> 5
 -- 5+ -> \x -> if
---     | x <= CDot           -> 5
+--     | x <= CDot           -> 3
```

Comment:
this is not correct. Or if you think it should be that way, then above regular expression construction is wrong (which I doubt). So most likely the Haskell `state5` implementation is also wrong.

```haskell
*Distribution.PackageDescription.Check> isGoodRelativeFilePath "foo./bar"
Just "same directory segment: .."
```

---

You can put the code above in the comment into Path.hs and

```
cabal-env kleene lattices
ghci Path.hs
*Main> Kleene.Internal.Pretty.putPretty $ fromTM filePathR
```

Reply:
Good catch, that's what I get for not diagramming things out.  State 3 is already reached via state 2, so this particular fix was exactly backwards.